# Problem Set 04: Numbers and NumPy - comments/solutions

# Programming for Economists Autumn 2026: Class 3
**Matias BF Hall**

**Follows Lecture 04**

This problem set covers floating-point comparisons, NumPy arrays, element-by-element calculations, shapes, axes, broadcasting, and Boolean indexing.

Use NumPy operations for the calculations. Do not replace array operations with loops. Plotting, pandas, optimization, and random numbers are not needed.

Before running each calculation, predict the shape of its result.

## 1. Floating-Point Numbers and Real Wages

The arrays below contain an average monthly wage in thousands of DKK and a consumer price index with 2021 as the base year. Three economists deflate the wage with expressions that are identical on paper:

- `nominal_wage / cpi * 100`
- `nominal_wage * (100 / cpi)`
- `100 * nominal_wage / cpi`

1. Calculate the three real-wage arrays.
2. Compare the first array with the second, and the first with the third, element by element using `==`. Count the disagreements in each Boolean array with `np.sum(...)`.
3. Print the years in which each exact comparison fails. Select them from `wage_years` with the Boolean array, and check the positions with `np.where(...)`.
4. Repeat both comparisons with `np.isclose(...)` and with `np.allclose(...)`.
5. Print the two difference arrays and state how large the differences are relative to the real wage level itself.
6. Round the three arrays to two decimals with `np.round(..., 2)` and compare them with `==` again.
7. Calculate the total percentage growth of the nominal wage and of the real wage between 2021 and 2025.

The two exact comparisons fail in different years. Explain in one sentence why, given that all three expressions describe the same economic quantity. Then add one sentence on what the two growth rates from step 7 say about this period.


In [26]:
import numpy as np

wage_years = np.array([2021, 2022, 2023, 2024, 2025])
nominal_wage = np.array([39.77, 40.40, 41.47, 43.76, 45.90])
cpi = np.array([100.0, 107.7, 111.3, 113.0, 115.0])

# 1.
real_wage_1 = nominal_wage / cpi * 100
real_wage_2 = nominal_wage * (100 / cpi)
real_wage_3 = 100 * nominal_wage / cpi

# 2.
bool_array1 = np.array([])
bool_array2 = np.array([])
for i, wage in enumerate(real_wage_1): # loop through each element in the array, returning the index and the value
    bool_array1 = np.append(bool_array1, wage == real_wage_2[i])
    bool_array2 = np.append(bool_array2, wage == real_wage_3[i])
# since we are using numpy arrays each array can be compared element-wise directly: bool_array1 = real_wage_1 == real_wage_2...

print('2.')
print(f'The number of disagreements between real_wage_1 and real_wage_2 is: {len(bool_array1) - np.sum(bool_array1)}')
print(f'The number of disagreements between real_wage_1 and real_wage_3 is: {len(bool_array2) - np.sum(bool_array2)}')
# as a true statement is equal to 1, and a false statement is equal to 0, we can sum the boolean array to get the number of true statements. We can then subtract this from the length of the array to get the number of false statements.
# we could also count by using negation. Usually in Python we write 'not' infront of a boolean statement. In numpy we can use the tilde ~ to negate a boolean arrays: np.sum(~bool_array1) 


# 3.
print('\n3.') # \n for newline
# we index with the boolean array. Numpy arrays can be indexed with boolean arrays, where True values indicate the elements to keep and False values indicate the elements to discard.
# bool_array1 == False will return a boolean array where it is True in the years that the comparison fails, and False in the years that the comparison is true.
print(f'The comparison between real_wage_1 and real_wage_2 is fails in the years: {wage_years[bool_array1 == False]}')
print(f'The comparison between real_wage_1 and real_wage_3 is fails in the years: {wage_years[bool_array2 == False]}')
# get positions using np.where
print(f'The comparison between real_wage_1 and real_wage_2 is fails in the positions: {np.where(bool_array1 == False)}') # again I use '.. == False' to negate the boolean array to be true where the comparison fails instead of where it succeds. We could us ~ instead
print(f'The comparison between real_wage_1 and real_wage_3 is fails in the positions: {np.where(bool_array2 == False)}') # np.where finds the indices of the array where the condition is true.

# 4.
print('\n4.')
print('isclose')
print(f'Using np.isclose the comparison between real_wage_1 and real_wage_2 is: {np.isclose(real_wage_1, real_wage_2)} and the comparison between real_wage_1 and real_wage_3 is: {np.isclose(real_wage_1, real_wage_3)}')
# np.isclose compares two arrays element-wise and returns a boolean array where True indicates that the elements are close to each other within a certain tolerance. The tolerance can be changed but default is fine.
print('allclose')
print(f'Using np.allclose the comparison between real_wage_1 and real_wage_2 is: {np.allclose(real_wage_1, real_wage_2)} and the comparison between real_wage_1 and real_wage_3 is: {np.allclose(real_wage_1, real_wage_3)}')
# np.allclose compares two arrays element-wise and returns a single boolean value indicating whether all elements are close to each other within a certain tolerance.
# use allclose to cheack all elements at once. If you want to find where they differ use is close and np.where/indexing to find where the differences are.

# 5.
print('\n5.')
diff_1_2 = np.abs(real_wage_1 - real_wage_2) # I use np.abs to the the absolute difference. This prevents negative values for the relative comparison.
diff_1_3 = np.abs(real_wage_1 - real_wage_3)
print(f'The absolute differences between real_wage_1 and real_wage_2 is: {diff_1_2}. Relative to the real wage the difference is: {diff_1_3/real_wage_1*100}%')
print(f'The absolute differences between real_wage_1 and real_wage_3 is: {diff_1_3}. Relative to the real wage the difference is: {diff_1_3/real_wage_1*100}%')
# the differences are tiny, all smaller than the 14th decimal place. The differences are due to the way floating point numbers are represented in computer memory

# 6.
print('\n6.')
bool_array1_round = np.round(real_wage_1, 2) == np.round(real_wage_2, 2)
bool_array2_round = np.round(real_wage_1, 2) == np.round(real_wage_3, 2)
print(f'When rounding to 2 decimal places the comparison between real_wage_1 and real_wage_2 fails {np.sum(~bool_array1_round)} times and the comparison between real_wage_1 and real_wage_3 fails {np.sum(~bool_array2_round)} times.')

# 7.
print('\n7.')
print(f'From 2021 to 2025 the real wage has increased by {real_wage_1[-1]/real_wage_1[0]-1:.2%} while the nominal wage has increased by {nominal_wage[-1]/nominal_wage[0]-1:.2%}')

2.
The number of disagreements between real_wage_1 and real_wage_2 is: 2.0
The number of disagreements between real_wage_1 and real_wage_3 is: 3.0

3.
The comparison between real_wage_1 and real_wage_2 is fails in the years: [2022 2024]
The comparison between real_wage_1 and real_wage_3 is fails in the years: [2022 2023 2024]
The comparison between real_wage_1 and real_wage_2 is fails in the positions: (array([1, 3]),)
The comparison between real_wage_1 and real_wage_3 is fails in the positions: (array([1, 2, 3]),)

4.
isclose
Using np.isclose the comparison between real_wage_1 and real_wage_2 is: [ True  True  True  True  True] and the comparison between real_wage_1 and real_wage_3 is: [ True  True  True  True  True]
allclose
Using np.allclose the comparison between real_wage_1 and real_wage_2 is: True and the comparison between real_wage_1 and real_wage_3 is: True

5.
The absolute differences between real_wage_1 and real_wage_2 is: [0.00000000e+00 7.10542736e-15 0.00000000e+00 7.1054

**Why the two exact comparisons fail in different years:**
The comparisons fail in different years due to the way floating-point arithmetic rounding errors accumulate differently in each expression. The comparisons fail in the years 2022, 2023 and 2024. Let's look at the differences in the expressions with numbers.  
2022: `real_wage_1` $ = 40.40 / 107.7 \cdot 100.0$, `real_wage_2` $ = 40.40 \cdot (100.0 / 107.7)$, and `real_wage_3` $ = 100.0 \cdot 40.40 / 107.7$. Since all numbers are floating-point numbers, the order of operations and the way the numbers are represented in memory can lead to slight differences in the results. This is why the comparisons fail in different years.

**What the two growth rates show:**  
Inflation crowds out the nominal wage growth, leaving little growth in the real wages.


<details>
<summary><strong>Hints and Results</strong></summary>

- All three real-wage arrays have shape `(5,)` and print as `[39.77       37.51160631 37.25965858 38.72566372 39.91304348]`.
- A Boolean array can be summed directly, because `True` counts as `1` and `False` as `0`.
- `real_wage_one == real_wage_two` gives `[True, False, True, False, True]`, so two comparisons fail: 2022 and 2024.
- `real_wage_one == real_wage_three` gives `[True, False, False, False, True]`, so three comparisons fail: 2022, 2023 and 2024.
- `np.where(...)` returns the positions inside a tuple: `(array([1, 3]),)` for the first comparison and `(array([1, 2, 3]),)` for the second.
- `np.isclose(...)` is `True` everywhere, and `np.allclose(...)` is `True` for both pairs.
- Every non-zero difference is `-7.10542736e-15`. That is about sixteen orders of magnitude below a real wage of roughly `38` thousand DKK.
- `np.round(array, 2)` is the function form of the method `array.round(2)` used in the lecture. After rounding, all three arrays are `[39.77 37.51 37.26 38.73 39.91]` and every `==` comparison is `True`.
- Between 2021 and 2025 the nominal wage grows by about `15.41` percent and the real wage by about `0.36` percent.

The three expressions divide and multiply in a different order, so the unavoidable rounding happens at a different step. Which years disagree is therefore a property of the arithmetic order, not of the data, and it moves when that order changes. Rounding to the precision an economic table actually reports removes the difference entirely.

</details>


## 2. A Household Consumption Basket

The arrays below describe one month of household consumption. Multiplying each price by its matching quantity gives spending in that category.

1. Print the shapes and data types of the three arrays.
2. Calculate spending by category using one element-by-element operation.
3. Calculate total monthly spending.
4. Calculate each category's percentage share of total spending.
5. Create a Boolean mask for categories with a budget share above 8 percent.
6. Use the mask to print the selected category names and their budget shares.

In one sentence, explain why element-by-element multiplication is appropriate here.

In [2]:
import numpy as np

consumption_categories = np.array(["Food", "Housing", "Energy", "Transport"])
prices = np.array([25.0, 8000.0, 4.0, 20.0])
quantities = np.array([30.0, 1.0, 250.0, 40.0])

In [7]:
# 1.
print('1.')
print(f'The shape of the category array is: {consumption_categories.shape}, the shape of the prices array is: {prices.shape} and the shape of the quantities array is: {quantities.shape}')
print(f'The data type of the category array is: {consumption_categories.dtype}, the data type of the prices array is: {prices.dtype} and the data type of the quantities array is: {quantities.dtype}')

# 2.
spending = prices * quantities

# 3.
total_monthly_spending = spending.sum()

# 4.
spending_share = spending / total_monthly_spending * 100

# 5.
I = spending_share > 8

# 6.
print('\n6.')
print(f'The categories with a spending share larger than 8% are: {consumption_categories[I]} and the corresponding spending shares are: {spending_share[I]}')

1.
The shape of the category array is: (4,), the shape of the prices array is: (4,) and the shape of the quantities array is: (4,)
The data type of the category array is: <U9, the data type of the prices array is: float64 and the data type of the quantities array is: float64

6.
The categories with a spending share larger than 8% are: ['Housing' 'Energy'] and the corresponding spending shares are: [75.82938389  9.47867299]


**Why element-by-element multiplication is appropriate:**

Because we have two arrays with the same shape, where each element in the arrays corresponds to a specific category(the same index in both arrays).

<details>
<summary><strong>Hints and Results</strong></summary>

- All three starting arrays have shape `(4,)`.
- `prices` and `quantities` have data type `float64`. `consumption_categories` has data type `<U9`, which means a string of at most nine characters, the length of `"Transport"`.
- The spending array is `[750.0, 8000.0, 1000.0, 800.0]` DKK.
- Total monthly spending is `10550.0` DKK.
- The budget shares are approximately `[7.11, 75.83, 9.48, 7.58]` percent.
- The mask for shares above 8 percent is `[False, True, True, False]`.
- The selected categories are `Housing` and `Energy`.

Each price and quantity at the same position describes the same consumption category, so the matching elements should be multiplied.

</details>

## 3. Real Wage Growth Across Countries and Years

Rows in `nominal_wage_growth` are countries and columns are years. The values are illustrative percentage growth rates. `inflation_by_year` contains one inflation rate for each year.

1. Print the shape of `nominal_wage_growth` and store its two dimensions as `n_countries` and `n_years`.
2. Select Sweden's complete row and the complete 2024 column.
3. Subtract `inflation_by_year` from the two-dimensional wage-growth array in one expression. Store the result as `real_wage_growth`.
4. Calculate average real wage growth for each country and for each year. Pay attention to the axis in each calculation.
5. Create a mask for countries whose average real wage growth exceeds 1.5 percent. Use it to print their names.
6. Create a mask for negative entries in `real_wage_growth`. Use `np.where(...)` to identify the corresponding row and column.

In one sentence, explain why the inflation array can be subtracted from every row without a loop.

In [8]:
import numpy as np

growth_years = np.array([2023, 2024, 2025])
country_names = np.array(["Denmark", "Sweden", "Germany", "Spain"])

nominal_wage_growth = np.array([
    [4.1, 3.6, 3.2],
    [4.5, 4.4, 3.7],
    [5.2, 4.8, 3.9],
    [2.4, 3.3, 2.9],
])

inflation_by_year = np.array([3.3, 2.1, 1.9])

In [30]:
# 1.
print('1.')
print(f'The shape of the nominal_wage_growth array is: {nominal_wage_growth.shape}')
n_countries, n_years = nominal_wage_growth.shape

# 2.
print('\n2.')
print(f'Sweden\'s row is: {nominal_wage_growth[country_names == "Sweden"]}, and the complete 2024 column is: {nominal_wage_growth[:, growth_years == 2024]}') # whe use \ before ' to espace the string formatting.
# we use boolean indexing to get index of Sweden and 2024 because it is more accurate, and would also work for longer arrays. Here it is perhaps easier to just count the index of Sweden and 2024.

# 3.
real_wage_growth = nominal_wage_growth - inflation_by_year
# since both arrays share at least one dimension, here the second, we can use broadcasting to subtract in one line.

# 4.
average_wage_growth_country = real_wage_growth.mean(axis=1) # with axis=1 we take the mean across rows. axis=.. chooses which axis to colapse. i.e. colapsing the second axis (axis=1) colapses the columns and returns a row vector with the mean of each row.
average_wage_growth_year = real_wage_growth.mean(axis=0) # with axis=0 we take the mean across columns.

# 5.
print('\n5.')
mask = average_wage_growth_country > 1.5
print(f'The countries with an average real wage growth larger than 1.5% are: {country_names[mask]}')

# 6.
negative_real_wage_growth = real_wage_growth < 0
negative_indexes = np.where(negative_real_wage_growth)
print('\n6.')
print(f'The countries and years with negative real wage growth are: {country_names[negative_indexes[0]]} in {growth_years[negative_indexes[1]]}')

1.
The shape of the nominal_wage_growth array is: (4, 3)

2.
Sweden's row is: [[4.5 4.4 3.7]], and the complete 2024 column is: [[3.6]
 [4.4]
 [4.8]
 [3.3]]

5.
The countries with an average real wage growth larger than 1.5% are: ['Sweden' 'Germany']

6.
The countries and years with negative real wage growth are: ['Spain'] in [2023]


In [26]:
average_wage_growth_year

array([1.2       , 1.76666667, 2.2       , 0.43333333])

**Why broadcasting works here:**

Since both arrays share at least one dimension, here the second, we can use broadcasting to subtract in one line.

<details>
<summary><strong>Hints and Results</strong></summary>

- The nominal wage-growth array has shape `(4, 3)`: four countries and three years.
- Sweden's row is `[4.5, 4.4, 3.7]`; the 2024 column is `[3.6, 4.4, 4.8, 3.3]`.
- `real_wage_growth` is

  `[[0.8, 1.5, 1.3], [1.2, 2.3, 1.8], [1.9, 2.7, 2.0], [-0.9, 1.2, 1.0]]`.
- Average real wage growth by country is approximately `[1.20, 1.77, 2.20, 0.43]`.
- Average real wage growth by year is `[0.75, 1.925, 1.525]`.
- The country mask is `[False, True, True, False]`, selecting Sweden and Germany.
- On a two-dimensional array `np.where(...)` returns a tuple of two arrays: first the row positions, then the matching column positions. Here it prints `(array([3]), array([0]))`.
- The only negative observation is in row `3`, column `0`: Spain in 2023, at `-0.9` percent.

The three inflation values match the three columns, so NumPy broadcasts the one-dimensional array across all four country rows.

</details>

## 4. Find and Fix the Errors

The rows below represent households and the columns represent two years. A policy simulation should add 2,000 DKK to first-year incomes below 20,000 DKK without changing the original data. It should then identify households that remain below 20,000 DKK and calculate average income in each year.

The code contains three errors. Copy it into the empty code cell below and run it. The first run stops with an error message instead of printing anything, so read that message first: it points straight at one of the three problems. Then correct one problem at a time and rerun after each change.

```python
import numpy as np

household_names = np.array(["Household A", "Household B", "Household C"])
income_by_year = np.array([
    [18000.0, 20000.0],
    [25000.0, 26000.0],
    [14000.0, 15000.0],
])

baseline_income = income_by_year[:, 0]
policy_income = baseline_income
policy_income[policy_income < 20000.0] += 2000.0

still_below_threshold = (policy_income > 0) and (policy_income < 20000.0)
average_income_by_year = np.mean(income_by_year, axis=1)

print("Baseline income:", baseline_income)
print("Income after support:", policy_income)
print("Still below the threshold:", household_names[still_below_threshold])
print("Average income by year:", average_income_by_year)
```

Write one sentence explaining each correction.

In [47]:
import numpy as np

household_names = np.array(["Household A", "Household B", "Household C"])
income_by_year = np.array([
    [18000.0, 20000.0],
    [25000.0, 26000.0],
    [14000.0, 15000.0],
])

baseline_income = income_by_year[:, 0] # Since we don't modify baseline_income, we can use a view of the original array. This is more memory efficient than creating a copy of the array.
policy_income = baseline_income.copy()  # Create a copy of baseline_income to avoid modifying it directly
policy_income[policy_income < 20000.0] += 2000.0

still_below_threshold = (policy_income > 0) & (policy_income < 20000.0) # use & for element-wise logical AND instead of 'and' with numpy arrays.
average_income_by_year = np.mean(income_by_year, axis=0) # use axis=0 to calculate the average income for each year.

print("Baseline income:", baseline_income)
print("Income after support:", policy_income)
print("Still below the threshold:", household_names[still_below_threshold])
print("Average income by year:", average_income_by_year)

Baseline income: [18000. 25000. 14000.]
Income after support: [20000. 25000. 16000.]
Still below the threshold: ['Household C']
Average income by year: [19000.         20333.33333333]


**Errors corrected:**

1. ***ValueError***: `and` is not defined for arrays and only expects a single Boolean value. Use `&` for element-wise logical AND instead of `and` with numpy arrays.
2. ***Wrong axis***: `axis=1` collapses the second dimension (columns) instead of the first (rows). Hence numpy sums and divides along rows, giving us one mean per row, i.e. the household means. Use `axis=0` to calculate the average income for each year.
3. ***Reference to same place in memory***: `policy_income = baseline_income` creates a reference to the same array in memory, so modifying `policy_income` also modifies `baseline_income`. Use `policy_income = baseline_income.copy()`. In this instance we even modify the original array, `income_by_year`, since `baseline_income = income_by_year[:, 0]`. This is due to the way Numpy handles slicing. When slicing lists, the result is a new list, but when slicing arrays, the result is a view of the original array. Hence, without copying `baseline_income`, `baseline_income`, `policy_income` andthe first column of `income_by_year`, all refer to the same place in memory, and the original data is changed.

<details>
<summary><strong>Hints and Results</strong></summary>

- A NumPy slice can be a view of the original array. The policy incomes need an independent `.copy()` before they are changed.
- Element-by-element logical comparisons use `&`, with each comparison inside parentheses. Python's `and` cannot compare two arrays and is what raises the error.
- `policy_income[policy_income < 20000.0] += 2000.0` assigns through a Boolean mask, changing only the selected positions. If the array on the left is a view, the original data changes with it.
- One average is needed for each year, so the household dimension must be removed by the correct axis choice.
- The unchanged first-year baseline is `[18000.0, 25000.0, 14000.0]`.
- Income after support is `[20000.0, 25000.0, 16000.0]`.
- Only `Household C` remains below 20,000 DKK.
- Average income by year is approximately `[19000.0, 20333.33]` DKK.

</details>